File: apps/ios/SafeWorldCore/Sources/SafeWorldCore/RemoteUpdate.swift
import Foundation

/// Thrown when a fetched payload is encoded in a way this app can't use.
public enum RemoteUpdateFormatError: LocalizedError {
    case hashedListForAndroid
    case unrecognized(String)

    public var errorDescription: String? {
        switch self {
        case .hashedListForAndroid:
            return "That list is hashed for Android and can't be used here — "
                + "point this app at the scrambled list instead."
        case .unrecognized(let format):
            return "Unrecognized list format \"\(format)\"."
        }
    }
}

/// Mirrors `packages/core/src/types.ts` (`RemoteUpdatePayload`). Domains are
/// keyed by the category's raw string id (e.g. "gambling") so the JSON shape
/// matches what the JS side of the project produces/consumes.
public struct RemoteUpdatePayload: Codable, Equatable {
    public var version: Int
    /// Identifies this exact set of domains, so a client can skip work it has already done. Absent on payloads published before the field existed.
    public var updateId: String?
    /// How `domains` is encoded — see `Scramble`. Absent means plaintext, so
    /// payloads published before this field keep working.
    public var format: String?
    public var domains: [String: [String]]

    public init(
        version: Int,
        updateId: String? = nil,
        format: String? = nil,
        domains: [String: [String]]
    ) {
        self.version = version
        self.updateId = updateId
        self.format = format
        self.domains = domains
    }

    /// `domains` re-keyed by `CategoryId` and decoded to plaintext, dropping any
    /// unrecognized category keys.
    ///
    /// Throws rather than returning an empty result for a format this app can't
    /// use: silently yielding nothing would leave the app looking healthy while
    /// blocking nothing new, which is the worst way for this to fail.
    public func domainsByCategory() throws -> [CategoryId: [String]] {
        if format == Scramble.hashedFormat { throw RemoteUpdateFormatError.hashedListForAndroid }
        if let format, format != Scramble.format {
            throw RemoteUpdateFormatError.unrecognized(format)
        }

        var result: [CategoryId: [String]] = [:]
        for (key, value) in domains {
            guard let id = CategoryId(rawValue: key) else { continue }
            result[id] = format == Scramble.format
                ? value.map(Scramble.unscramble).filter { !$0.isEmpty }
                : value
        }
        return result
    }
}

File: apps/windows/SafeWorld.Core/RemoteUpdate.cs
namespace SafeWorld.Core;

/// <summary>Thrown when a fetched payload is encoded in a way this app can't use.</summary>
public sealed class RemoteUpdateFormatException : Exception
{
    public RemoteUpdateFormatException(string message) : base(message)
    {
    }

    public static RemoteUpdateFormatException HashedListForAndroid() => new(
        "That list is hashed for Android and can't be used here — " +
        "point this app at the scrambled list instead.");

    public static RemoteUpdateFormatException Unrecognized(string format) => new(
        $"Unrecognized list format \"{format}\".");
}

/// <summary>
/// Mirrors <c>packages/core/src/types.ts</c> (<c>RemoteUpdatePayload</c>). Domains are keyed by
/// the category's raw string id (e.g. "list2") so the JSON shape matches what the JS side of the
/// project produces/consumes.
/// </summary>
public sealed class RemoteUpdatePayload
{
    public int Version { get; set; }

    /// <summary>Identifies this exact set of domains, so a client can skip work it has already done. Absent on payloads published before the field existed.</summary>
    public string? UpdateId { get; set; }

    /// <summary>How <see cref="Domains"/> is encoded — see <see cref="Scramble"/>. Null means plaintext, so payloads published before this field keep working.</summary>
    public string? Format { get; set; }

    public Dictionary<string, List<string>> Domains { get; set; } = new();

    /// <summary>
    /// <see cref="Domains"/> re-keyed by <see cref="CategoryId"/> and decoded to plaintext,
    /// dropping any unrecognized category keys.
    ///
    /// Throws rather than returning an empty result for a format this app can't use: silently
    /// yielding nothing would leave the app looking healthy while blocking nothing new, which is
    /// the worst way for this to fail.
    /// </summary>
    public Dictionary<CategoryId, List<string>> DomainsByCategory()
    {
        if (Format == Scramble.HashedFormat) throw RemoteUpdateFormatException.HashedListForAndroid();
        if (Format is not null && Format != Scramble.Format) throw RemoteUpdateFormatException.Unrecognized(Format);

        var result = new Dictionary<CategoryId, List<string>>();
        foreach (var (key, value) in Domains)
        {
            var id = CategoryIdExtensions.FromRawValue(key);
            if (id is null) continue;
            result[id.Value] = Format == Scramble.Format
                ? value.Select(Scramble.UnscrambleDomain).Where(d => d.Length > 0).ToList()
                : value;
        }
        return result;
    }
}
